cmd/ipfs-cluster-service/main.go

```diff
@@ -632,3 +644,19 @@ func getStateManager() cmdutils.StateManager {
 	checkErr("creating state manager", err)
 	return mgr
 }
+
+func getRandomPortMultiaddr(assigned map[int]bool) ma.Multiaddr {
```

Comment:
Rather than passing-in a map, pass in multiaddresses and:

1) Iterate using https://godoc.org/github.com/multiformats/go-multiaddr#ForEach
2) When the `Protocol()` is TCP or UDP, obtain a random port.
3) Replace the component (which is passed by value) with a new one using the random port
3) As the ForEach progresses, re-`Encapsulate() the mulitaddress into a new one.
4) When all the multiaddresses are finished, close the listeners.
(This is just a way that I see of doing it consistently for any multiaddress regardless of the form. If passing in pointers to multiaddresses, you can update the pointer destinations before returning to the new addresses, without needing to assign the returned values).